File: 2/js/src/Powerset.ts
function generatePowerset(list: number[]): number[][] {
	const powerset: number[][] = [];
	powerset.push([]);

	for(const num of list) {
		const sets: number[][] = [];

		const length = powerset.length;
		for(let i=0; i<length; i++) {
			const set = powerset[i];
			powerset.push(set.concat(num));
		}
	}

	return powerset;
}

(function main() {
	const list = [1,2,3];
	console.log(generatePowerset(list));
})();

File: 2/java/Powerset.java
import java.util.*;

class Powerset {
	/*	
	public static List<Object> getPowerset(List<Integer> set) {
		List<Object> powerset = new ArrayList<Object>();
	
		int selected = 0;

		while(selected <= Math.pow(2,set.size())-1)	{
			String binaryNumber = Integer.toBinaryString(selected); 
			List<Integer> subset = new ArrayList<Integer>();	

			for(int i=binaryNumber.length()-1; i>=0; i--) {
				if(binaryNumber.charAt(i) == '1') {
					subset.add(set.get(binaryNumber.length()-i-1));
				}
			}		
			powerset.add(subset);
			selected = selected + 1;
		}

		return powerset;
	}	
	*/
	
	public static List<List<Integer>> getPowerset(List<Integer> set) {
		List<List<Integer>> powerset = new ArrayList<>();	
		powerset.add(Arrays.asList(new Integer[]{}));
		
		for(Integer num : set) {
			int n = powerset.size();
			for(int i=0; i<n; i++) {
				List<Integer> subset = new ArrayList<>(powerset.get(i));
				subset.add(num);
				powerset.add(subset);
			}
		}	 


		return powerset;
	}

	public static void main(String args[]) {
		List<Integer> set = new ArrayList<>();
		set = Arrays.asList(new Integer[]{1,2,3});

		System.out.println(getPowerset(set));
	}
}
